data/transactions/logic/assembler.go

```diff
@@ -0,0 +1,1053 @@
+		}
+		ops.Out.WriteByte(0x27) // bytec
+		ops.Out.WriteByte(uint8(constIndex))
+	}
+	ops.tpush(StackBytes)
+	return nil
+}
+
+// ByteLiteral writes opcodes and data for loading a []byte literal
+// Values are accumulated so that they can be put into a bytecblock
+func (ops *OpStream) ByteLiteral(val []byte) error {
+	found := false
+	var constIndex uint
+	for i, cv := range ops.bytec {
+		if bytes.Compare(cv, val) == 0 {
+			found = true
+			constIndex = uint(i)
+			break
+		}
+	}
+	if !found {
+		constIndex = uint(len(ops.bytec))
+		ops.bytec = append(ops.bytec, val)
+	}
+	return ops.Bytec(constIndex)
+}
+
+// Arg writes opcodes for loading from Lsig.Args
+func (ops *OpStream) Arg(val uint64) error {
+	switch val {
+	case 0:
+		ops.Out.WriteByte(0x2d) // arg_0
+	case 1:
+		ops.Out.WriteByte(0x2e) // arg_1
+	case 2:
+		ops.Out.WriteByte(0x2f) // arg_2
+	case 3:
+		ops.Out.WriteByte(0x30) // arg_3
+	default:
+		if val > 0xff {
+			return errors.New("cannot have more than 256 args")
+		}
+		ops.Out.WriteByte(0x2c)
+		ops.Out.WriteByte(uint8(val))
+	}
+	ops.tpush(StackBytes)
+	return nil
+}
+
+// Txn writes opcodes for loading a field from the current transaction
+func (ops *OpStream) Txn(val uint64) error {
+	if val >= uint64(len(TxnFieldNames)) {
+		return errors.New("invalid txn field")
+	}
+	ops.Out.WriteByte(0x31)
+	ops.Out.WriteByte(uint8(val))
+	ops.tpush(TxnFieldTypes[val])
+	return nil
+}
+
+// Gtxn writes opcodes for loading a field from the current transaction
+func (ops *OpStream) Gtxn(gid, val uint64) error {
+	if val >= uint64(len(TxnFieldNames)) {
+		return errors.New("invalid txn field")
+	}
+	if gid > 255 {
+		return errors.New("gtxn cannot look up beyond group index 255")
+	}
+	ops.Out.WriteByte(0x33)
+	ops.Out.WriteByte(uint8(gid))
+	ops.Out.WriteByte(uint8(val))
+	ops.tpush(TxnFieldTypes[val])
+	return nil
+}
+
+// Global writes opcodes for loading an evaluator-global field
+func (ops *OpStream) Global(val uint64) error {
+	if val >= uint64(len(GlobalFieldNames)) {
+		return errors.New("invalid txn field")
+	}
+	ops.Out.WriteByte(0x32)
+	ops.Out.WriteByte(uint8(val))
+	ops.tpush(GlobalFieldTypes[val])
+	return nil
+}
+
+func assembleInt(ops *OpStream, args []string) error {
+	val, err := strconv.ParseUint(args[0], 0, 64)
+	if err != nil {
+		return err
+	}
+	return ops.Uint(val)
+}
+
+// Explicit invocation of const lookup and push
+func assembleIntC(ops *OpStream, args []string) error {
+	constIndex, err := strconv.ParseUint(args[0], 0, 64)
+	if err != nil {
+		return err
+	}
+	return ops.Intc(uint(constIndex))
+}
+func assembleByteC(ops *OpStream, args []string) error {
+	constIndex, err := strconv.ParseUint(args[0], 0, 64)
+	if err != nil {
+		return err
+	}
+	return ops.Bytec(uint(constIndex))
+}
+
+func parseBinaryArgs(args []string) (val []byte, consumed int, err error) {
+	arg := args[0]
+	if strings.HasPrefix(arg, "base32(") || strings.HasPrefix(arg, "b32(") {
```

Comment:
Feature request: `string "asdf"` puts `[]byte("asdf")` into the program.

Reply:
Yes, except string escape sequences are annoying to parse and it's Friday and I'm tired. It's a good idea though. Maybe next week (-: